src/pounder/mod.rs

```diff
@@ -0,0 +1,186 @@
+        let result = match mode {
+            ad9959::Mode::SingleBitTwoWire | ad9959::Mode::SingleBitThreeWire =>
+                self.qspi.configure_mode(hal::qspi::QspiMode::OneBit),
+            ad9959::Mode::TwoBitSerial => self.qspi.configure_mode(hal::qspi::QspiMode::TwoBit),
+            ad9959::Mode::FourBitSerial => self.qspi.configure_mode(hal::qspi::QspiMode::FourBit),
+        };
+
+        result.map_err(|_| Error::Qspi)
+    }
+
+    fn write(&mut self, addr: u8, data: &[u8]) -> Result<(), Error> {
+        if (addr & 0x80) != 0 {
+            return Err(Error::InvalidAddress);
+        }
+
+        self.qspi.write(addr, &data).map_err(|_| Error::Qspi)
+    }
+
+    fn read(&mut self, addr: u8, mut dest: &mut [u8]) -> Result<(), Error> {
+        if (addr & 0x80) != 0 {
+            return Err(Error::InvalidAddress);
+        }
+        self.qspi.read(0x80_u8 | addr, &mut dest).map_err(|_| Error::Qspi)
+    }
+}
+
+pub struct PounderDevices<DELAY> {
+    pub ad9959: ad9959::Ad9959<QspiInterface,
+                               DELAY,
+                               hal::gpio::gpiog::PG7<hal::gpio::Output<hal::gpio::PushPull>>>,
+    mcp23017: mcp23017::MCP23017<hal::i2c::I2c<hal::stm32::I2C1>>,
+    attenuator_spi: hal::spi::Spi<hal::stm32::SPI1>,
+    adc1: hal::adc::Adc<hal::stm32::ADC1, hal::adc::Enabled>,
+    adc2: hal::adc::Adc<hal::stm32::ADC2, hal::adc::Enabled>,
+    adc1_in_p: hal::gpio::gpiof::PF11<hal::gpio::Analog>,
+    adc2_in_p: hal::gpio::gpiof::PF14<hal::gpio::Analog>,
+}
+
+impl<DELAY> PounderDevices<DELAY>
+where
+    DELAY: embedded_hal::blocking::delay::DelayMs<u8>,
+{
+    pub fn new(mcp23017: mcp23017::MCP23017<hal::i2c::I2c<hal::stm32::I2C1>>,
+               ad9959: ad9959::Ad9959<QspiInterface,
+                                      DELAY,
+                                      hal::gpio::gpiog::PG7<
+                                        hal::gpio::Output<hal::gpio::PushPull>>>,
+               attenuator_spi: hal::spi::Spi<hal::stm32::SPI1>,
+               adc1: hal::adc::Adc<hal::stm32::ADC1, hal::adc::Enabled>,
+               adc2: hal::adc::Adc<hal::stm32::ADC2, hal::adc::Enabled>,
+               adc1_in_p: hal::gpio::gpiof::PF11<hal::gpio::Analog>,
+               adc2_in_p: hal::gpio::gpiof::PF14<hal::gpio::Analog>,
+               ) -> Result<Self, Error> {
+        let mut devices = Self {
+            mcp23017,
+            ad9959,
+            attenuator_spi,
+            adc1,
+            adc2,
+            adc1_in_p,
+            adc2_in_p,
+        };
+
+        // Configure power-on-default state for pounder. All LEDs are on, on-board oscillator
+        // selected, attenuators out of reset.
+        devices.mcp23017.write_gpioa(0xF).map_err(|_| Error::I2c)?;
+        devices.mcp23017.write_gpiob(1_u8.wrapping_shl(5)).map_err(|_| Error::I2c)?;
+        devices.mcp23017.all_pin_mode(mcp23017::PinMode::OUTPUT).map_err(|_| Error::I2c)?;
```

Comment:
The mcp23017 crate really needs a couple of fixes.
I think [`all_pin_mode`](https://github.com/lucazulian/mcp23017/blob/09ff49d6b7ffd20d21d31430917aac922c130b05/src/lib.rs#L164) is broken. It set's just the LSB.
And `digital_write` and `_read` appear to be badly defined.
Also https://github.com/lucazulian/mcp23017/issues

Reply:
I agree that the crate is a bit clunky and has a few bugs present. I'll fork the repository and fix some bugs (and try to contribute them back to mainline).

Reply:
There's actually a PR already that cleans things up: https://github.com/mrd0ll4r/mcp23017
I'm changing our dependency to point to that PR for the time being.